cuda : Add conv2d Implicit GEMM

bssrdf opened the pull request:
This PR added another CUDA conv_2d op using implicit GEMM approach. It is only optimized for cuda cores and its performance is up to 10x of that of direct method currently in llama.cpp.
 
On a RTX4090
|Cases | Direct | Implicit GEMM |
|---|---|---|
| ne_input=[19,19,256,16],ne_kernel=[4,4,256,4096], |    2.23 TFLOPS  |   38.76 TFLOPS |
| ne_input=[19,19,8,16],ne_kernel=[4,4,8,128], |    1.85 TFLOPS  |    9.12 TFLOPS |
| ne_input=[19,19,8,16],ne_kernel=[4,4,8,130], |    1.76 TFLOPS  |    9.27 TFLOPS |
| ne_input=[19,19,4,16],ne_kernel=[2,2,4,4], |  147.71 GFLOPS  |  150.00 GFLOPS |
| ne_input=[224,224,3,1],ne_kernel=[3,3,3,8], |    1.04 TFLOPS  |    1.02 TFLOPS  |
| ne_input=[224,224,1,1],ne_kernel=[2,2,1,8], |  255.40 GFLOPS  |  238.21 GFLOPS |
| ne_input=[224,224,1,8],ne_kernel=[2,2,1,8], |  308.44 GFLOPS |  324.17 GFLOPS |
| ne_input=[58,58,32,1],ne_kernel=[3,3,32,64], |    1.49 TFLOPS  |    3.98 TFLOPS |
| ne_input=[58,58,32,8],ne_kernel=[3,3,32,64], |    1.88 TFLOPS  |   15.85 TFLOPS |
| ne_input=[16,16,128,8],ne_kernel=[3,3,128,512], |    1.98 TFLOPS  |   16.90 TFLOPS |
| ne_input=[19,19,256,16],ne_kernel=[4,4,256,4096], |    2.27 TFLOPS  |   38.00 TFLOPS |
| ne_input=[19,19,8,16],ne_kernel=[4,4,8,128], |    1.86 TFLOPS |    8.64 TFLOPS |
| ne_input=[19,19,8,16],ne_kernel=[4,4,8,130], |    1.80 TFLOPS  |    8.78 TFLOPS |
| ne_input=[19,19,4,16],ne_kernel=[2,2,4,4], |  150.12 GFLOPS  |  147.95 GFLOPS  |
| ne_input=[224,224,3,1],ne_kernel=[3,3,3,8], |    1.01 TFLOPS  |  980.39 GFLOPS |
| ne_input=[224,224,1,1],ne_kernel=[2,2,1,8], |  245.83 GFLOPS  |  212.52 GFLOPS |
| ne_input=[224,224,1,8],ne_kernel=[2,2,1,8], |  305.41 GFLOPS  |  317.95 GFLOPS |
| ne_input=[58,58,32,1],ne_kernel=[3,3,32,64], |    1.43 TFLOPS  |    3.74 TFLOPS |
| ne_input=[58,58,32,8],ne_kernel=[3,3,32,64], |    1.81 TFLOPS  |   14.96 TFLOPS |
| ne_input=[16,16,128,8],ne_kernel=[3,3,128,512], |    1.84 TFLOPS  |   15.80 TFLOPS |

Comparison with im2col+gemm

Fp16 filter, Fp32 activation
| (IC, OC, IW, IH) | im2col+GEMM TIME | im2col+GEMM VRAM |  implicit GEMM TIME | implicit GEMM VRAM
| --- | --- | --- |  --- | ---
 | (64, 64, 48, 64) | 0.03 ms | 4.12 MB | 0.07 ms | 0.75 MB
 | (320, 320, 104, 152) | 0.56 ms | 106.13 MB | 0.98 ms | 19.30 MB
 | (640, 640, 52, 76) | 0.32 ms | 53.07 MB | 1.24 ms | 9.65 MB
 | (640, 640, 104, 152) | 1.41 ms | 212.27 MB | 3.04 ms | 38.59 MB
 | (960, 320, 104, 152) | 1.48 ms | 279.80 MB | 2.68 ms | 19.30 MB
 | (1280, 1280, 26, 38) | 0.21 ms | 26.53 MB | 1.19 ms | 4.82 MB
 | (1280, 640, 52, 76) | 0.62 ms | 96.48 MB | 2.33 ms | 9.65 MB
 | (1920, 1280, 26, 38) | 0.30 ms | 37.39 MB | 1.79 ms | 4.82 MB
 | (2560, 1280, 26, 38) | 0.42 ms | 48.24 MB | 2.36 ms | 4.82 MB
 | (512, 512, 104, 152) | 0.91 ms | 169.81 MB | 1.88 ms | 30.88 MB
 | (512, 512, 208, 304) | 3.90 ms | 679.25 MB | 7.95 ms | 123.50 MB
 | (512, 256, 416, 608) | 12.55 ms | 2470.00 MB | 15.67 ms | 247.00 MB
 | (256, 128, 832, 1216) | 24.82 ms | 4940.00 MB | 15.67 ms | 494.00 MB
 | (256, 256, 832, 1216) | 27.43 ms | 5434.00 MB | 31.17 ms | 988.00 MB
 | (320, 256, 1024, 1920) | 66.56 ms | 12720.00 MB | 76.05 ms | 1920.00 MB

Fp32 filter, Fp32 activation
| (IC, OC, IW, IH) | im2col+GEMM TIME | im2col+GEMM VRAM |  implicit GEMM TIME | implicit GEMM VRAM
| --- | --- | --- |  --- | ---
 | (64, 64, 48, 64) | 0.04 ms | 7.50 MB | 0.07 ms | 0.75 MB
 | (320, 320, 104, 152) | 0.92 ms | 192.97 MB | 0.90 ms | 19.30 MB
 | (640, 640, 52, 76) | 0.68 ms | 96.48 MB | 1.19 ms | 9.65 MB
 | (640, 640, 104, 152) | 2.41 ms | 385.94 MB | 2.95 ms | 38.59 MB
 | (960, 320, 104, 152) | 2.38 ms | 540.31 MB | 2.56 ms | 19.30 MB
 | (1280, 1280, 26, 38) | 0.71 ms | 48.24 MB | 1.10 ms | 4.82 MB
 | (1280, 640, 52, 76) | 1.18 ms | 183.32 MB | 2.20 ms | 9.65 MB
 | (1920, 1280, 26, 38) | 0.72 ms | 69.95 MB | 1.83 ms | 4.82 MB
 | (2560, 1280, 26, 38) | 0.94 ms | 91.66 MB | 2.35 ms | 4.82 MB
 | (512, 512, 104, 152) | 1.57 ms | 308.75 MB | 1.79 ms | 30.88 MB
 | (512, 512, 208, 304) | 6.34 ms | 1235.00 MB | 7.61 ms | 123.50 MB
 | (512, 256, 416, 608) | 17.49 ms | 4693.00 MB | 15.00 ms | 247.00 MB
 | (256, 128, 832, 1216) | 32.16 ms | 9386.00 MB | 15.06 ms | 494.00 MB
 | (256, 256, 832, 1216) | 36.54 ms | 9880.00 MB | 30.23 ms | 988.00 MB
 | (320, 256, 1024, 1920) | 562.36 ms | 23520.00 MB | 73.56 ms | 1920.00 MB

bssrdf committed:
Add implicit GEMM convolution operation for 2D tensors in CUDA

bssrdf committed:
Add implicit convolution support for 2D tensors in CPU and CUDA implementations

bssrdf committed:
fix passing param as reference

bssrdf committed:
Fix parameter order in conv2d_implicit and add comprehensive test cases for 2D convolution

bssrdf committed:
Fix boundary check in conv2d_implicit_kernel to include channel limits

bssrdf commented:
> Why are you adding a new ggml op?

Because of https://github.com/ggml-org/llama.cpp/pull/15669#discussion_r2311865862

leejet commented:
I think the implementation of implicit gemm can directly use ggml_conv2d_direct. There's really no need to provide so many conv2d functions.

bssrdf commented:
> I think the implementation of implicit gemm can directly use ggml_conv2d_direct. There's really no need to provide so many conv2d functions.

I can reuse ggml_conv2d_direct. TBH it is not a very good or intuitive name (the best one, ggml_conv_2d, is already occupied). I do wish it has an additional argument (ggml_conv_2d should carry in the beginning) for what method implemented.

JohannesGaessler commented:
What should be done regarding `IM2COL` vs. `CONV2D` is to construct the compute graph using `CONV2D` and to then let each backend decide how to do the operation. If a backend lacks support for convolution it should allocate a temporary buffer for `IM2COL` and use that as a workaround.

For kernel selection, please take a look at how e.g. `FLASH_ATTN_EXT` is being handled. There are multiple kernels that can be used, at runtime one is selected based on hardware capabilities and tensor shapes. All convolution kernels that do the exact same operation should be using the same ggml op. If we have multiple kernels that could be used we need to test which code paths are faster under which circumstances and write the logic accordingly. This is particularly relevant because there is a concurrent PR using tensor cores: https://github.com/ggml-org/llama.cpp/pull/15813 . cc @mnehete32

JohannesGaessler reviewed (commented):
For this PR, try removing the current conv2d kernel and replacing it with this one. Chances are it will be universally faster since it uses shared memory and has (unless I misread the code) coalesced memory accesses. I'll test the performance using a P40, RTX 3090, and RTX 4090 for NVIDIA and an RX 6800 and Mi 50 for AMD.

bssrdf commented on ggml/src/ggml-cuda/conv2d-implicit.cu:
```diff
@@ -0,0 +1,365 @@
+#include "convert.cuh"
+
+typedef struct{    
+    unsigned int      n;                              //batch szie              
```
Done

bssrdf commented on ggml/src/ggml-cuda/conv2d-implicit.cu:
```diff
@@ -0,0 +1,365 @@
+
+typedef struct{    
+    unsigned int      n;                              //batch szie              
+    unsigned int      c;                              //channel number          
```
done

bssrdf commented on ggml/src/ggml-cuda/conv2d-implicit.cu:
```diff
@@ -0,0 +1,365 @@
+    int posh_ori[4];
+    int posw_ori[4];
+#pragma unroll
+    for (int i = 0; i < 4; ++i)
+    {
```
Done. corrected styles in all places

JohannesGaessler replied:
Speaking of P40s, you should be careful with FP16 arithmetic since that is massively gimped on Pascal. You can use the macro `FAST_FP16_AVAILABLE` to check whether FP16 would be fast and use FP32 as a workaround if not. You can look at e.g. `mmvf.cu` for an example.

bssrdf committed:
Refactor conv2d_implicit_kernel for improved readability and consistency; update parameter comments and remove unused code

bssrdf committed:
Refactor conv2d_implicit_kernel for improved bitwise operations; add test for implicit convolution

bssrdf committed:
merged with upstream master

bssrdf committed:
Merge branch 'master' into conv2d-implicit

etasnadi commented:
> > Corrected test drive in sd.cpp for VAE decoding:
> > ### 768x1024 sd1 fp16 vae:
> > method	time	memory
> > CUDA imcol+mul	~1.68s	4992.19 MB
> > CUDA direct	~35.35s	1920.19 MB
> > CUDA implicitgemm	~2.20s	1920.19 MB
> > VULKAN imcol+mul	OOM	~4992 MB
> > VULKAN direct	~1.17s	1920.19 MB
> > Ignore previous numbers. Now looking more like it. Good job @bssrdf :) , sorry for the confusion.
> > edit: added vulkan numbers from same device.
> 
> Thanks for the update, @Green-Sky. The sd.cpp result is consistent with what I tested in ggml backend op. Surprised to see VULKAN is so fast 🤔. Curious to know what optimization it used. A pity I don't know how to program in it.
> 
> I hope to improve implicit's performance, especially for Fp16. There should be plenty of low hanging fruits there.

Hi, there are a few optimizations used in Vulkan: mininizing branch divergence, reducing the number of modulo/div computations with shuffles. Later @jeffbolznv further optimized it by using constant int divs and optimized the blocktile sizes to common inputs.

Although the GLSL code is not too hard to read, I added the CUDA translation of the GLSL shader in a new PR: https://github.com/ggml-org/llama.cpp/pull/15805. It also introduces some bank conflict reduction, and vectorized shmem loads so it is now faster than Vulkan on large inputs.

bssrdf commented:
> > > Corrected test drive in sd.cpp for VAE decoding:
> > > ### 768x1024 sd1 fp16 vae:
> > > method	time	memory
> > > CUDA imcol+mul	~1.68s	4992.19 MB
> > > CUDA direct	~35.35s	1920.19 MB
> > > CUDA implicitgemm	~2.20s	1920.19 MB
> > > VULKAN imcol+mul	OOM	~4992 MB
> > > VULKAN direct	~1.17s	1920.19 MB
> > > Ignore previous numbers. Now looking more like it. Good job @bssrdf :) , sorry for the confusion.
> > > edit: added vulkan numbers from same device.
> > 
> > 
> > Thanks for the update, @Green-Sky. The sd.cpp result is consistent with what I tested in ggml backend op. Surprised to see VULKAN is so fast 🤔. Curious to know what optimization it used. A pity I don't know how to program in it.
> > I hope to improve implicit's performance, especially for Fp16. There should be plenty of low hanging fruits there.
> 
> Hi, there are a few optimizations used in Vulkan: mininizing branch divergence, reducing the number of modulo/div computations with shuffles. Later @jeffbolznv further optimized it by using constant int divs and optimized the blocktile sizes to common inputs.
> 
> Although the GLSL code is not too hard to read, I added the CUDA translation of the GLSL shader in a new PR: #16088. It also introduces some bank conflict reduction, and vectorized shmem loads so it is now faster than Vulkan on large inputs.

@etasnadi, thanks for working on the cuda backend. Yes, with these optimizations in place, cuda should be on par with vulkan if not faster. Great job!

bssrdf committed:
minor update and add direct conv in benchmarking

etasnadi commented:
> Now #16088 is a better implementation.

Actually, it is a somewhat better implementation you just forget to enable it in your code so the previous inefficient implementation was used, so I suggest to further merge on this PR.

See my comment in: https://github.com/etasnadi/llama.cppxx/pull/1#issuecomment-3341613849

bssrdf committed:
minor updates

bssrdf committed:
reorder register tile loop

bssrdf committed:
work in progress

bssrdf committed:
added block variants; to be debugged

bssrdf committed:
fixed missing dilation

bssrdf committed:
add support for both NCHW and NHWC layouts

bssrdf committed:
unroll some loops

bssrdf committed:
change padding size to 1; added padding to input smem

bssrdf committed:
change padding size back to 4

bssrdf committed:
turn on tests for implicit conv2d

bssrdf committed:
WIP

Commit:
WIP

Commit:
WIP

bssrdf committed:
WIP

bssrdf committed:
WIP

bssrdf committed:
WIP: output

bssrdf committed:
WIP: tensore code compiled ok

bssrdf committed:
WIP: debugging tensor core kernel

bssrdf committed:
WIP: bug fix

bssrdf committed:
WIP: fixed another bug

bssrdf committed:
properly use __CUDA_ARCH__ to protect the tensor path

bssrdf committed:
this case is broken; to be debugged

bssrdf committed:
still debugging

bssrdf committed:
WIP: bug fix

bssrdf committed:
WIP: fixed another bug

bssrdf committed:
WIP: cleanup; enhanced test case

bssrdf committed:
WIP

Commit:
added a specialization for cuda copy op when tensor is transposed

bssrdf committed:
WIP: debugging cpy transpose

Commit:
WIP: fixed a bug in cpy transpos index computation

bssrdf committed:
WIP: almost working

bssrdf committed:
WIP: minor tweak

bssrdf commented:
> I'm not seeing a copy kernel in this PR, do you have it somewhere else? In any case, I think that optimizing the copy kernel would best be spun out into a separate PR since it can be reviewed and merged independently of the convolution code.

Thanks, @JohannesGaessler. The copy kernel is on the tensor path branch https://github.com/bssrdf/llama.cpp/tree/improve-cuda-cpy-op.

bssrdf committed:
reduced bank conflicts for output

bssrdf committed:
switch to default conv2d interface

bssrdf committed:
clean up

bssrdf commented:
Please review this PR and ignore the changes related to cpy (they will be reverted back; I have another PR addressing it).

The updated numbers on 4090
```
 CONV_2D_IMPLICIT(ne_input=[19,19,256,16],ne_kernel=[4,4,256,4096],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                   249 runs -  4022.96 us/run - 137.42 GFLOP/run -  34.16 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,8,16],ne_kernel=[4,4,8,128],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              46376 runs -    21.86 us/run - 133.69 MFLOP/run -   6.12 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,8,16],ne_kernel=[4,4,8,130],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              45694 runs -    22.15 us/run - 135.78 MFLOP/run -   6.13 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,4,16],ne_kernel=[2,2,4,4],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               163840 runs -     6.26 us/run - 642.82 kFLOP/run - 102.69 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,3,1],ne_kernel=[3,3,3,8],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               43074 runs -    25.17 us/run -  20.90 MFLOP/run - 830.24 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,1,1],ne_kernel=[2,2,1,8],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               73728 runs -    14.80 us/run -   2.78 MFLOP/run - 188.12 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,1,8],ne_kernel=[2,2,1,8],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               13467 runs -    84.18 us/run -  22.28 MFLOP/run - 264.64 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[58,58,32,1],ne_kernel=[3,3,32,64],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              25143 runs -    39.84 us/run - 115.40 MFLOP/run -   2.90 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[58,58,32,8],ne_kernel=[3,3,32,64],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              13734 runs -    72.93 us/run - 923.24 MFLOP/run -  12.66 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[16,16,128,8],ne_kernel=[3,3,128,512],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                    6160 runs -   163.35 us/run -   1.85 GFLOP/run -  11.32 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,256,16],ne_kernel=[4,4,256,4096],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                  1059 runs -   945.04 us/run - 137.42 GFLOP/run - 145.41 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,8,16],ne_kernel=[4,4,8,128],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              34408 runs -    29.17 us/run - 133.69 MFLOP/run -   4.58 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,8,16],ne_kernel=[4,4,8,130],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              34639 runs -    29.09 us/run - 135.78 MFLOP/run -   4.67 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[19,19,4,16],ne_kernel=[2,2,4,4],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               171864 runs -     6.09 us/run - 642.82 kFLOP/run - 105.60 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,3,1],ne_kernel=[3,3,3,8],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               38288 runs -    27.32 us/run -  20.90 MFLOP/run - 764.80 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,1,1],ne_kernel=[2,2,1,8],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               73656 runs -    15.14 us/run -   2.78 MFLOP/run - 183.92 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[224,224,1,8],ne_kernel=[2,2,1,8],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               13467 runs -    85.62 us/run -  22.28 MFLOP/run - 260.21 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[58,58,32,1],ne_kernel=[3,3,32,64],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              25143 runs -    40.52 us/run - 115.40 MFLOP/run -   2.85 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[58,58,32,8],ne_kernel=[3,3,32,64],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):              20492 runs -    49.00 us/run - 923.24 MFLOP/run -  18.84 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[16,16,128,8],ne_kernel=[3,3,128,512],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                    8635 runs -   115.98 us/run -   1.85 GFLOP/run -  15.94 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,512,1],ne_kernel=[3,3,512,512],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   2038 runs -   491.00 us/run -  57.98 GFLOP/run - 118.08 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[192,256,512,1],ne_kernel=[3,3,512,512],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   537 runs -  1865.09 us/run - 231.90 GFLOP/run - 124.34 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,256,1],ne_kernel=[3,3,256,256],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   385 runs -  2601.56 us/run - 231.88 GFLOP/run -  89.13 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,128,1],ne_kernel=[3,3,128,128],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                  194 runs -  5165.32 us/run - 231.83 GFLOP/run -  44.88 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,512,1],ne_kernel=[1,1,512,512],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                   4368 runs -   229.79 us/run -   6.44 GFLOP/run -  28.01 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,4,1],ne_kernel=[1,1,4,4],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               188416 runs -     5.43 us/run - 344.06 kFLOP/run -  63.35 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,4,1],ne_kernel=[3,3,4,512],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):              36512 runs -    27.54 us/run - 446.69 MFLOP/run -  16.22 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[3,3,512,512],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   133 runs -  7565.57 us/run - 927.61 GFLOP/run - 122.61 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[3,3,512,256],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   200 runs -  5015.85 us/run - 463.81 GFLOP/run -  92.47 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[1,1,512,256],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                   568 runs -  1762.81 us/run -  51.49 GFLOP/run -  29.21 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[3,3,256,128],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                  107 runs -  9365.56 us/run - 463.76 GFLOP/run -  49.52 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[1,1,256,128],type_kernel=f16,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                  538 runs -  1859.06 us/run -  51.44 GFLOP/run -  27.67 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[3,3,256,256],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   96 runs - 10423.55 us/run - 927.51 GFLOP/run -  88.98 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,128,1],ne_kernel=[3,3,128,3],type_kernel=f16,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                    323 runs -  3138.93 us/run -   5.43 GFLOP/run -   1.73 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,512,1],ne_kernel=[3,3,512,512],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                    570 runs -  1755.68 us/run -  57.98 GFLOP/run -  33.02 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[192,256,512,1],ne_kernel=[3,3,512,512],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   149 runs -  6718.10 us/run - 231.90 GFLOP/run -  34.52 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,256,1],ne_kernel=[3,3,256,256],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   149 runs -  6756.52 us/run - 231.88 GFLOP/run -  34.32 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,128,1],ne_kernel=[3,3,128,128],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                  151 runs -  6662.15 us/run - 231.83 GFLOP/run -  34.80 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,512,1],ne_kernel=[1,1,512,512],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                   4912 runs -   204.09 us/run -   6.44 GFLOP/run -  31.54 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,4,1],ne_kernel=[1,1,4,4],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):               188416 runs -     5.34 us/run - 344.06 kFLOP/run -  64.46 GFLOPS
  CONV_2D_IMPLICIT(ne_input=[96,128,4,1],ne_kernel=[3,3,4,512],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):              35616 runs -    28.22 us/run - 446.69 MFLOP/run -  15.83 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[3,3,512,512],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                    39 runs - 26276.90 us/run - 927.61 GFLOP/run -  35.30 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[3,3,512,256],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                    75 runs - 13360.11 us/run - 463.81 GFLOP/run -  34.72 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[384,512,512,1],ne_kernel=[1,1,512,256],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                   628 runs -  1592.81 us/run -  51.49 GFLOP/run -  32.33 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[3,3,256,128],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   75 runs - 13406.59 us/run - 463.76 GFLOP/run -  34.59 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[1,1,256,128],type_kernel=f32,stride0=1,stride1=1,padding0=0,padding1=0,dilation0=1,dilation1=1,cwhn=0):                  588 runs -  1701.20 us/run -  51.44 GFLOP/run -  30.24 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,256,1],ne_kernel=[3,3,256,256],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                   38 runs - 26836.76 us/run - 927.51 GFLOP/run -  34.56 TFLOPS
  CONV_2D_IMPLICIT(ne_input=[768,1024,128,1],ne_kernel=[3,3,128,3],type_kernel=f32,stride0=1,stride1=1,padding0=1,padding1=1,dilation0=1,dilation1=1,cwhn=0):                    171 runs -  6131.79 us/run -   5.43 GFLOP/run - 886.11 GFLOPS
```

bssrdf committed:
remove implicit op and related calls; replace conv_2d with conv_2d_implicit kernel

bssrdf committed:
clean up

bssrdf committed:
more clean up

bssrdf commented on ggml/include/ggml.h:
```diff
@@ -1982,6 +1983,19 @@ extern "C" {
             int                   d0,  // dilation dimension 0
             int                   d1); // dilation dimension 1
 
+    GGML_API struct ggml_tensor * ggml_conv_2d_implicitgemm(
```
Done! I replaced ```conv_2d_direct``` with ```conv_2d_implicit``` for ```CONV_2D``` op in ```ggml_cuda.c```. Please give it a try in sd.cpp. I hope there is no a quality degradation.

bssrdf committed:
prevent CI compile failure

bssrdf committed:
make CI happy

bssrdf committed:
make CI happy

bssrdf committed:
make CI happy

bssrdf commented:
> Re did the sd.cpp testing with all new commits (sd+ggml+pr), also heat soaked and **_more tests_** and sampling performance.
> 
> ### sd1 fp16 512x768
> method	time sample	memory sampling	time decoding	memory decoding
> CUDA imcol+mul	0.21s	189.38 MB	0.75s	2496.09 MB
> CUDA direct (master)	2.96s	132.71 MB	16.79s	1056.09 MB
> **CUDA direct (bssrdf pr)**	0.37s	132.71 MB	1.00s	1056.09 MB
> ### sd1 fp16 768x1024 (like the old table)
> method	time sample	memory sampling	time decoding	memory decoding
> CUDA imcol+mul	0.58s	373.64 MB	1.55s	4992.19 MB
> CUDA direct (master)	6.29s	260.30 MB	34.94s	2112.19 MB
> **CUDA direct (bssrdf pr)**	0.85s	260.30 MB	2.03s	2112.19 MB
> ### sdxl fp16/q8_0 1024x1280
> Diffusion model is q8_0 and vae is fp16.
> 
> method	time sample	memory sampling	time decoding	memory decoding
> CUDA imcol+mul	0.79s	614.83 MB	OOM	9600.31 MiB (alloc error)
> CUDA direct (master)	11.05s	288.43 MB	60.57s	4800.31 MB
> **CUDA direct (bssrdf pr)**	1.15s	288.43 MB	3.72s	4800.31 MB
> "sample/sampling" -> diffusion step "decoding" -> VAE decoding
> 
> Sampling time normalized (multiplied/divided by 2, because cfg != 1). All runs with flash attention in the diffusion model.
> 
> ```
> [INFO ] stable-diffusion.cpp:285  - Version: SD 1.x
> [INFO ] stable-diffusion.cpp:312  - Weight type stat:                      f16: 1131
> [INFO ] stable-diffusion.cpp:313  - Conditioner weight type stat:          f16: 196
> [INFO ] stable-diffusion.cpp:314  - Diffusion model weight type stat:      f16: 686
> [INFO ] stable-diffusion.cpp:315  - VAE weight type stat:                  f16: 248
> ```
> 
> ```
> [INFO ] stable-diffusion.cpp:285  - Version: SDXL
> [INFO ] stable-diffusion.cpp:312  - Weight type stat:                      f32: 1185 |     f16: 124  |    q8_0: 1332
> [INFO ] stable-diffusion.cpp:313  - Conditioner weight type stat:          f32: 354  |    q8_0: 359
> [INFO ] stable-diffusion.cpp:314  - Diffusion model weight type stat:      f32: 707  |    q8_0: 973
> [INFO ] stable-diffusion.cpp:315  - VAE weight type stat:                  f32: 124  |     f16: 124
> ```

@Green-Sky,  thank you for the extensive test. Appreciated. So this PR is still 25% slower. 
BTW, which GPU are you using?

bssrdf committed:
added a test case to directly compare im2col and implicit gemm

bssrdf committed:
make tensor core path available for cc 7.5 and above

bssrdf commented:
> ```
> $ result/bin/test-conv2d
> ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
> ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
> ggml_cuda_init: found 1 CUDA devices:
>   Device 0: NVIDIA GeForce RTX 2070, compute capability 7.5, VMM: yes
> ```
> 
> (IC, OC, IW, IH)	im2col+GEMM TIME	im2col+GEMM VRAM	implicit GEMM TIME	implicit GEMM VRAM
> (320, 1280, 26, 38, 3, 3)	0.70 ms	10.25 MB	0.64 ms	4.82 MB
> This test is very very fast and short, and the variance is relatively high.

Interesting. On 4090, im2col is consistently faster for small input sizes

```
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4090, compute capability 8.9, VMM: yes
```

| (IC, OC, IW, IH, KW, KH) | im2col+GEMM TIME | im2col+GEMM VRAM | implicit GEMM TIME | implicit GEMM VRAM
| --- | --- | --- | --- | ---
|  (320, 1280, 26, 38, 3, 3) | 0.14 ms | 10.25 MB | 0.26 ms | 4.82 MB 
| (64, 64, 48, 64, 3, 3) | 0.03 ms | 4.12 MB | 0.07 ms | 0.75 MB
 | (320, 320, 104, 152, 3, 3) | 0.59 ms | 106.13 MB | 0.28 ms | 19.30 MB
 | (640, 640, 52, 76, 3, 3) | 0.38 ms | 53.07 MB | 0.51 ms | 9.65 MB
 | (640, 640, 104, 152, 3, 3) | 1.49 ms | 212.27 MB | 1.10 ms | 38.59 MB
 | (960, 320, 104, 152, 3, 3) | 1.64 ms | 279.80 MB | 0.86 ms | 19.30 MB
 | (1280, 1280, 26, 38, 3, 3) | 0.25 ms | 26.53 MB | 1.02 ms | 4.82 MB
 | (1280, 1280, 26, 38, 1, 1) | 0.09 ms | 13.67 MB | 0.19 ms | 5.47 MB
 | (256, 128, 768, 1024, 3, 3) | 21.93 ms | 3840.00 MB | 7.06 ms | 384.00 MB
 | (128, 3, 768, 1024, 3, 3) | 10.26 ms | 1737.00 MB | 3.00 ms | 9.00 MB
 | (256, 128, 768, 1024, 1, 1) | 8.78 ms | 771.50 MB | 1.83 ms | 385.75 MB
 | (512, 256, 384, 512, 1, 1) | 3.78 ms | 387.51 MB | 1.73 ms | 193.75 MB
 | (1280, 640, 52, 76, 3, 3) | 0.72 ms | 96.48 MB | 0.99 ms | 9.65 MB
 | (1920, 1280, 26, 38, 3, 3) | 0.37 ms | 37.39 MB | 1.63 ms | 4.82 MB
 | (2560, 1280, 26, 38, 3, 3) | 0.51 ms | 48.24 MB | 2.17 ms | 4.82 MB
 | (320, 1280, 26, 38, 3, 3) | 0.13 ms | 10.25 MB | 0.27 ms | 4.82 MB
 | (512, 512, 104, 152, 3, 3) | 1.04 ms | 169.81 MB | 0.48 ms | 30.88 MB
 | (512, 512, 208, 304, 3, 3) | 4.32 ms | 679.25 MB | 1.96 ms | 123.50 MB
 | (512, 256, 416, 608, 3, 3) | 13.98 ms | 2470.00 MB | 4.84 ms | 247.00 MB
 | (256, 128, 832, 1216, 3, 3) | 27.65 ms | 4940.00 MB | 9.07 ms | 494.00 MB
 | (256, 256, 832, 1216, 3, 3) | 30.35 ms | 5434.00 MB | 10.36 ms | 988.00 MB

bssrdf committed:
minor update of test case

bssrdf committed:
fixes for CI

bssrdf committed:
remove trailing blank

bssrdf committed:
another CI fix

bssrdf committed:
add more sd like test cases

bssrdf committed:
added split-k mode for skinny mnk shapes

bssrdf committed:
fixed bug now split-k is working

bssrdf committed:
use a better criterian to use split-k

bssrdf committed:
try to reduce index calculation

bssrdf committed:
refactor cuda core code path

bssrdf committed:
make sure there are enough channels for split-k

bssrdf committed:
Merge branch 'refactor-cuda-core-path' into conv2d-implicit

bssrdf committed:
added test cases exactly replicating sdxl unet steps

bssrdf committed:
minor update

bssrdf committed:
trying to get rid of remaining bank conflicts; also fixed a bug for split-k condition check

Commit:
WIP: debugging

bssrdf committed:
swizzling working, may still have room to optimize

bssrdf committed:
now bank conflicts free and performance get a bit boosted too

bssrdf committed:
further reduce index swizzling computation cycles

bssrdf committed:
remove commented code

bssrdf committed:
broken for some test cases

bssrdf committed:
fixed a bug now split-k can choose a better split factor

bssrdf committed:
make split-k condition check more robust

bssrdf committed:
make CI happy

bssrdf committed:
make CI happy

bssrdf committed:
reduced uncoalesced global access in filter transpose

bssrdf committed:
reduce bank conflicts in filter transpose

bssrdf committed:
add loop unrolling

bssrdf committed:
increase maximum split factor to 16; use better heuristics to choose split-K factor, reducing tail effect

bssrdf committed:
get rid of a convert unary kernel call and fuse the type cast into conv epilogue

bssrdf committed:
add missing batch index bounds check

bssrdf committed:
m16n8k16 mma works; to be cleaned up

bssrdf committed:
trying to reduce integer ops; simply code

bssrdf committed:
only enable m16n8k16 on ampere or above

bssrdf committed:
change mac loop to match cutlass

Commit:
restore smem pointer at teh end of evry rs loop

bssrdf committed:
WIP: move rs loop into block-k-loop following cutlass

bssrdf committed:
WIP: debugging

bssrdf committed:
WIP: fixed bugs now results are correct

bssrdf committed:
move some register to const memory space

bssrdf committed:
disable m16n8k16 mma for ampere for now

bssrdf committed:
reducing integer ops

bssrdf committed:
various small optimizations

Commit:
add/fix GGML_UNUSED

bssrdf committed:
WIP: adding cp.async calls

bssrdf committed:
WIP: debugging

bssrdf committed:
add the missing guard

bssrdf committed:
WIP

bssrdf committed:
not working properly for channel numbers of 32, 48, 96 etc., ok for 64, 128...

bssrdf committed:
the special filter transpose NCHW2NHWC is broken, disable it and use the other less optimized one

bssrdf committed:
restore split-k for small inputs

bssrdf committed:
fixed a bug in  the special filter transpose NCHW2NHWC; still failing for channel number < 32

bssrdf committed:
fixed anotehr bug in the special filter transpose NCHW2NHWC

bssrdf committed:
added a few edge test cases

bssrdf committed:
due to cp.async, only support filter size <= 32

bssrdf committed:
remove some repeated index computation; various code/comments clean up

bssrdf committed:
minor tweak filter tranpose

bssrdf committed:
tuned block dimensions for filter tranpose

bssrdf committed:
use fastdiv in filter transpose

bssrdf committed:
make CI happy

Commit:
remove trailing blanks

bssrdf committed:
further reduce repeated index comutations

bssrdf committed:
fix CI

bssrdf committed:
Merge branch 'master' into conv2d-implicit

bssrdf committed:
use existing cvta_generic_to_shared

bssrdf committed:
remove m16n8k16 path which is not faster than m16n8k8

bssrdf committed:
added a test case that failed in test-backend-ops

bssrdf committed:
turn on run whole graph to pass all test cases

jeffbolznv commented:
Actually, looks like the infs occur when the output is entirely zero, and the large values occur when the values are all tiny. In both cases, these are clearly incorrect and not a bug in the tests.

Also interesting that these seem to not happen when a test is run alone.

bssrdf commented:
> Then it would also need to cache and compare the node types. I'm kind of surprised it doesn't do this already.

Let me describe what happened in this PR when cuda graph is on. It is very interesting.
```
void ggml_cuda_op_conv2d_implicit(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
....
....

  if (kernel->type == GGML_TYPE_F16) {
        const half *             K_D    = (const half *) kernel->data;
        printf("Using conv2d f16 implicit GEMM\n");
        conv2d_implicit_cuda_f16(ctx, X_D, K_D, Y_D, cc, params, st);
    } else {
        const float *             K_D   = (const float *) kernel->data;
        printf("Using conv2d f32 implicit GEMM\n");
        conv2d_implicit_cuda_f32(ctx, X_D, K_D, Y_D, cc, params, st);
    }
}

```
This is the cuda op function/entry point. based on kernel tensor's type, it dispatches to two different paths. I put two prints to indicate which path it follows.  For the following two cases:
```
 test_cases.emplace_back(new test_conv_2d( { 1, 133, 1, 2}, { 1, 1, 1, 1},
         GGML_TYPE_F32, 1, 5, 5, 2, 2, 4, false));
 test_cases.emplace_back(new test_conv_2d( { 1, 133, 1, 2}, { 1, 1, 1, 1},
         GGML_TYPE_F16, 1, 5, 5, 2, 2, 4, false));
```
Only one print (fp32) showed up. For the second case, the `ggml_cuda_op_conv2d_implicit` function never executed but somehow the cuda kernel ran.

bssrdf committed:
Merge branch 'master' into conv2d-implicit

bssrdf committed:
remove run_whole_graph(); not needed after the fix

bssrdf commented:
Now all test cases passed.

bssrdf committed:
remove test-conv2d

bssrdf committed:
fix to work for Turing

bssrdf committed:
exclude HIP from using tensor core

Commit:
Merge branch 'master' into conv2d-implicit

bssrdf commented:
> So, how are we looking? How is HIP doing?

I don't know how to resolve the HIP build error. Anyone who could lend a hand will be appreciated.

daniandtheweb commented:
I've been able to fix the HIP build issue.

Apparently the compiler doesn't read `cudaFuncSetAttribute` at lines 967 and 1165 of conv2d-implicit.cu as it should and instead of seeing 3 arguments it sees everything as an argument, even the elements between brackets.

My solution has been to add a `(const void*)` cast to `conv2d_implicit_kernel...` .

By fixing this I've been able to compile the program and run it without any issue.

bssrdf committed:
use CUDA_SET_SHARED_MEMORY_LIMIT as a fix for HIp build

bssrdf committed:
Merge branch 'master' into conv2d-implicit

zhang-hui-yulo commented:
May I ask how to compile conv2d-implicit on HIP? I just checkout [bssrdf:conv2d-implicit](https://github.com/bssrdf/llama.cpp/tree/conv2d-implicit) and don't see compile error on my hip windows 6.4.2.

Based on my knowledge, hipFuncSetAttribute needs const void* as function type, so CUDA_SET_SHARED_MEMORY_LIMIT needs to be modified to support hip, or you can just get rid of the macro and directly call cudaFuncSetAttribute like  [fattn-mma-f16.cuh](https://github.com/ggml-org/llama.cpp/blob/7a99dc85e2d26a9b9c540cad887322eae8924f03/ggml/src/ggml-cuda/fattn-mma-f16.cuh#L1757)

bssrdf commented:
> May I ask how to compile conv2d-implicit on HIP? I just checkout [bssrdf:conv2d-implicit](https://github.com/bssrdf/llama.cpp/tree/conv2d-implicit) and don't see compile error on my hip windows 6.4.2.
> 
> Based on my knowledge, hipFuncSetAttribute needs const void* as function type, so CUDA_SET_SHARED_MEMORY_LIMIT needs to be modified to support hip, or you can just get rid of the macro and directly call cudaFuncSetAttribute like [fattn-mma-f16.cuh](https://github.com/ggml-org/llama.cpp/blob/7a99dc85e2d26a9b9c540cad887322eae8924f03/ggml/src/ggml-cuda/fattn-mma-f16.cuh#L1757)



> May I ask how to compile conv2d-implicit on HIP? I just checkout [bssrdf:conv2d-implicit](https://github.com/bssrdf/llama.cpp/tree/conv2d-implicit) and don't see compile error on my hip windows 6.4.2.
> 
> Based on my knowledge, hipFuncSetAttribute needs const void* as function type, so CUDA_SET_SHARED_MEMORY_LIMIT needs to be modified to support hip, or you can just get rid of the macro and directly call cudaFuncSetAttribute like [fattn-mma-f16.cuh](https://github.com/ggml-org/llama.cpp/blob/7a99dc85e2d26a9b9c540cad887322eae8924f03/ggml/src/ggml-cuda/fattn-mma-f16.cuh#L1757)

Hi, if there is no compile error, can you try it with HIP backends and see if it works? The tensor core path is not meant for non-NVIDIA devices but the other path using cuda core may work. Sorry, I don't know much about HIP.

zhang-hui-yulo commented:
> > I just look at the main loop of implicit gemm, too many native ptx code in it which makes the kernel only works in cuda, I will suggest to change the code to adapter mma.cuh or it's hard to be ported to other platorms.
> 
> Unfortunately I didn't use the existing mma tiles when developing the conv2d kernel. I am also not the right person to adapt this kernel to AMD devices due to knowledge gap. I think it is probably better to merge this PR and leave development for other platforms in future PRs. Thanks.

You still need to adapter other NVIDIA GPUs like Volta, anyway I'm not the moderator, using mma.cuh is just a suggestion, the decision shall belongs to @JohannesGaessler
